src/elexmodel/handlers/data/Featurizer.py

```diff
@@ -0,0 +1,123 @@
+import numpy as np
+import pandas as pd
+
+
+class Featurizer(object):
+    """
+    Featurizer. Normalizes features, add intercept, expands fixed effects
+    """
+
+    def __init__(self, features, fixed_effects):
+        self.features = features
+        self.fixed_effects = fixed_effects
+        self.expanded_fixed_effects = []
+        self.complete_features = None
+        self.column_means = None
+
+    def compute_means_for_centering(self, *arg):
+        """
+        Computes and saves the column mean of pandas dataframe passed as args.
+        This is used for centering.
+        """
+        data = pd.concat(arg)
+        self.column_means = data[self.features].mean()
+
+    def _center_features(self, df):
+        """
+        Centers the features. This changes the interpretation of the intercept coefficient
+        from conditional mean given covariates = 0, to conditional mean given covariates are
+        their average value
+        """
+        df[self.features] = df[self.features] - self.column_means
+
+    def _add_intercept(self, df):
+        df["intercept"] = 1
+
+    def _expand_fixed_effects(self, df, drop_first):
+        """
+        Convert fixed effect columns into dummy variables.
+        """
+        # we concatenate the dummy variables with the original fixed effects, since we need the original fixed
+        # effect columns for aggregation.
+        original_fixed_effect_columns = df[self.fixed_effects]
+        return pd.concat(
+            [
+                pd.get_dummies(
+                    df,
+                    columns=self.fixed_effects,
+                    prefix=self.fixed_effects,
+                    prefix_sep="_",
+                    dtype=np.int64,
+                    drop_first=drop_first,
+                ),
+                original_fixed_effect_columns,
+            ],
+            axis=1,
+        )
+
+    def featurize_fitting_data(self, fitting_data, center_features=True, add_intercept=True):
+        """
+        Featurize the data that the model is fitted on.
+        """
+        # make copy of fitting_data, since we do not want to change the original data
+        new_fitting_data = fitting_data.copy()
+        self.center_features = center_features
+        self.add_intercept = add_intercept
+
+        if self.center_features:
+            self._center_features(new_fitting_data)
+
+        if self.add_intercept:
+            self._add_intercept(new_fitting_data)
+
+        if len(self.fixed_effects) > 0:
+            # drop_first is True for fitting_data (reporting_units) since we want to avoid the design matrix with
```

Comment:
Can you clarify here in this comment all the places we have "fitting data" (I think its just reporting data and the training data df in the conform step?). Also we should write somewhere which fitting data dfs and prediction dfs have the same features, have a subset of each others features etc.

Reply:
ok, just added that to the functions docstrings, let me know what you think.
Also, write out which dataframes have the same features (and subsets etc.) when invoking the featurizer in `BaseElectionModel`

Reply:
Cool -- I think we can then drop the reference to  "(reporting units)" in line 74 or just say e.g. reporting units. Sorry to be anal.

Reply:
no worries and done!